File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        int minWindowSize = Integer.MAX_VALUE;
        int minStartIdx = 0;

        Map<Character, Integer> subStrMap = new HashMap<>();
        Map<Character, Integer> originStrMap = new HashMap<>();

        for(char ch : t.toCharArray()) {
            subStrMap.put(ch, subStrMap.getOrDefault(ch, 0) + 1);
        }

        int left = 0, right = 0, need = subStrMap.size(), have = 0;
        while(right < s.length()){
            char rightChar = s.charAt(right);
            originStrMap.put(rightChar, originStrMap.getOrDefault(rightChar, 0) + 1);

            if(subStrMap.containsKey(rightChar) &&
                    subStrMap.get(rightChar).intValue() == originStrMap.get(rightChar).intValue()) {
                have++;
            }

            while(need == have) {
                if(minWindowSize > right - left + 1) {
                    minWindowSize = right - left + 1;
                    minStartIdx = left;
                }

                char leftChar = s.charAt(left);
                originStrMap.put(leftChar, originStrMap.get(leftChar) - 1);

                if(subStrMap.containsKey(leftChar) &&
                        originStrMap.get(leftChar) < subStrMap.get(leftChar)) {
                    have--;
                }


                left++;
            }

            right++;
        }

        return minWindowSize == Integer.MAX_VALUE ? "" : s.substring(minStartIdx, minStartIdx + minWindowSize);
    }
}

File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        need = collections.Counter(t)
        missing = len(t)
        left = 0

        answer = ""

        for right in range(len(s)):
            char = s[right]

            if need[char] > 0:
                missing -= 1

            need[char] -= 1

            while missing == 0:
                window = s[left:right + 1]

                if answer == "" or len(window) < len(answer):
                    answer = window

                left_char = s[left]

                need[left_char] += 1

                if need[left_char] > 0:
                    missing += 1

                left += 1

        return answer